pkg/envinfo/url.go

```diff
@@ -0,0 +1,275 @@
+		if len(url.Container) == 0 {
+			url.Container = containerName
+		}
+	}
+
+	// container is not provided, or the specified port is not being used under any containers
+	// pick the first container to store the new endpoint
+	if len(url.Container) == 0 {
+		url.Container = containerComponents[0].Name
+	}
+
+	if url.Port == -1 {
+		var err error
+		url.Port, err = util.GetValidPortNumber(ei.GetName(), url.Port, ei.GetPorts())
+		if err != nil {
+			return err
+		}
+	}
+
+	// get the name
+	if len(url.Name) == 0 {
+		url.Name = util.GetURLName(ei.GetName(), url.Port)
+	}
+
+	return nil
+}
+
+// ValidateURL validates the given URL
+func (ei *EnvInfo) ValidateURL(url localConfigProvider.LocalURL) error {
+
+	containers, err := generator.GetContainers(ei.devfileObj)
+	if err != nil {
+		return err
+	}
+	if len(containers) == 0 {
+		return fmt.Errorf("no valid components found in the devfile")
+	}
+
+	foundContainer := false
+	containerComponents := ei.devfileObj.Data.GetDevfileContainerComponents()
+
+	// map TargetPort with containerName
+	containerPortMap := make(map[int]string)
+	for _, component := range containerComponents {
+		if len(url.Container) > 0 && !foundContainer {
+			if component.Name == url.Container {
+				foundContainer = true
+			}
+		}
+		for _, endpoint := range component.Container.Endpoints {
+			if endpoint.Name == url.Name {
+				return fmt.Errorf("url %v already exist in devfile endpoint entry under container %v", url.Name, component.Name)
+			}
+			containerPortMap[endpoint.TargetPort] = component.Name
+		}
+	}
+
+	if len(url.Container) > 0 && !foundContainer {
+		return fmt.Errorf("the container specified: %v does not exist in devfile", url.Container)
+	}
+	if containerName, exist := containerPortMap[url.Port]; exist {
+		if len(url.Container) > 0 && url.Container != containerName {
+			return fmt.Errorf("cannot set URL %v under container %v, TargetPort %v is being used under container %v", url.Name, url.Container, url.Port, containerName)
+		}
+	}
+
+	errorList := make([]string, 0)
+	if url.TLSSecret != "" && (url.Kind != localConfigProvider.INGRESS || !url.Secure) {
+		errorList = append(errorList, "TLS secret is only available for secure URLs of Ingress kind")
+	}
+	// check if a host is provided for route based URLs
+	if len(url.Host) > 0 {
+		if url.Kind == localConfigProvider.ROUTE {
+			errorList = append(errorList, "host is not supported for URLs of Route Kind")
+		}
+		if err := validation.ValidateHost(url.Host); err != nil {
+			errorList = append(errorList, err.Error())
+		}
+	} else if url.Kind == localConfigProvider.INGRESS {
+		errorList = append(errorList, "host must be provided in order to create URLS of Ingress Kind")
+	}
+	if len(url.Protocol) > 0 && (strings.ToLower(url.Protocol) != string(devfilev1.HTTPEndpointProtocol) && strings.ToLower(url.Protocol) != string(devfilev1.HTTPSEndpointProtocol) && strings.ToLower(url.Protocol) != string(devfilev1.WSEndpointProtocol) &&
```

Comment:
Can we use `switch` instead of `if` to make this more readable/maintainable?

Reply:
Fixed